Improve unsupported value handling in bolt messages

ali-ince opened the pull request:
Previously, the connection was terminated when an unsupported message / type was encountered during message deserialization. Although a couple of improvements were made over time, those were only handling very special cases, namely when `null` or `duplicate` keys were received for map values which was resulting in a `FAILURE` message.

While this PR mainly targets improving unsupported value handling, it contains following notable changes;

1. Splitting handshake & protocol logic into distinct netty handlers and enabling message deserialization to work on a one-message-containing-buffer without catering for unconsumed buffers on failures (we can leave unconsumed buffers in case of errors without worrying about integrity of upcoming messages),
2. Making `PackOutput` a part of `BoltConnection` which enables us to couple `flushing of output buffers` with actual `message handling` without any indirections.
3. Cleanup of complex `buffer` / `chunk` handling code,
4. Adding a distinction between pack stream errors that would cause `connection termination` and `FAILURE` message;
a. Incoming well known cypher value types cause `FAILURE` message to be generated with a status of `Status.Statement.TypeError` when not supported by the protocol version in place,
b. `Null` valued keys cause `FAILURE` message to be generated with a status of `Status.Request.Invalid`,
c. Duplicate keys cause `FAILURE` message to be generated with a status of `Status.Request.Invalid`,
d. Unkown `struct` / `message` types or unsupported key types for `Map`s will still trigger a connection termination,
e. All other errors that occurs during message deserialization will still trigger a connection termination.

ali-ince committed:
Improve inbound message handling and related cleanup

ali-ince committed:
Fail on unsupported types instead of terminating connections

zhenlineo commented on community/bolt/src/main/java/org/neo4j/bolt/messaging/KnownType.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public enum KnownType
```
Should this be versioned too as the "known types" to PackStream V1 and V2 are different?

And also as we have this type definitions, then maybe we could remove the duplicated const all over the places?
Such as in [`NeoPackV1`](https://github.com/neo4j/neo4j/blob/3.4/community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/Neo4jPackV1.java#L63)
```
    public static final byte NODE = 'N';
    public static final byte RELATIONSHIP = 'R';
    public static final byte UNBOUND_RELATIONSHIP = 'r';
    public static final byte PATH = 'P';
```

ali-ince replied:
I intentionally did not version this enum as there's a need for looking up all server supported types independent of its version. The version specific logic will still live in `PackerV1` and `PackerV2`.

Yep, that's right. I spent some time around removing those constants but removing them will either require an invocation of `KnownType#signature()` or a lookup through `KnownType#valueOf()` - which would cause the packing / unpacking logic to involve more code when compared to using constants. WDYT?

zhenlineo replied:
The main concern for me is that we defined this magic `N`, `R` etc. twice.
How about `NODE( NeoPackV1.Node, "Node" )`? Will this be simple and clear to read?

ali-ince replied:
👍I think so.

zhenlineo commented on community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/BoltIOException.java:
```diff
@@ -43,4 +43,9 @@ public Status status()
         return status;
     }
+
+    public boolean causesFailure()
```
causeFailure? or casualFailure, or causalFailure? maybe `isInvalidFormatError`?

zhenlineo commented on community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/BoltRequestMessageReader.java:
```diff
@@ -89,20 +80,20 @@ public void read( BoltRequestMessageHandler handler ) throws IOException
-                    throw new BoltIOException( Status.Request.Invalid,
-                            "Message 0x" + Integer.toHexString( signature ) + " is not supported." );
+                    throw new BoltIOException( Status.Request.InvalidFormat,
+                            String.format( "Message 0x%s is not supported.", Integer.toHexString( signature ) ) );
```
These error messages are not very helpful for us when a user come to us with such an error in their stack trace. We usually have to ask what the user have done. So maybe it would be useful and possible after this PR that we could print out the whole message that failed to decode. So that we could get a clue what did the user send.

This is a good-to-have but not need-to-fix in this PR thing.

ali-ince replied:
I'm carding this.

zhenlineo reviewed (commented):
Feels the `KnownType` should be versioned.

ali-ince committed:
Address review comments and report errors during type unpacking

ali-ince commented:
@zhenlineo thanks for the comments. 

I think I've addressed all of your comments plus I've also added support for communicating back errors that occur during unpacking of struct values (point dimension mismatch, invalid time zone name, unknown crsid, etc.).

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/messaging/KnownType.java:
```diff
@@ -0,0 +1,87 @@
+
+import static java.util.Collections.unmodifiableMap;
+
+public enum KnownType
```
Maybe `StructType`?

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/messaging/KnownType.java:
```diff
@@ -0,0 +1,87 @@
+{
+    NODE( Neo4jPackV1.NODE, "Node" ),
+    RELATIONSHIP( Neo4jPackV1.RELATIONSHIP, "Relationship" ),
+    UNBOUND_RELATIONSHIP( Neo4jPackV1.UNBOUND_RELATIONSHIP, "Relationship" ),
```
"UnboundRelationship" instead of "Relationship"?

ali-ince replied:
Previously `Relationship` was being used to log errors on `UNBOUND_RELATIONSHIP`, so I tried to conserve that behaviour. But I agree that having underlying type names reported is a better way to go.

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/messaging/KnownType.java:
```diff
@@ -0,0 +1,87 @@
+    UNBOUND_RELATIONSHIP( Neo4jPackV1.UNBOUND_RELATIONSHIP, "Relationship" ),
+    PATH( Neo4jPackV1.PATH, "Path" ),
+    POINT_2D( Neo4jPackV2.POINT_2D, "Point" ),
+    POINT_3D( Neo4jPackV2.POINT_3D, "Point" ),
```
I think it might be better to differentiate between 2D and 3D points since they are different structs in Bolt

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/transport/pipeline/MessageAccumulator.java:
```diff
@@ -0,0 +1,81 @@
+{
+    private boolean readMessageBoundary;
+
+    public MessageAccumulator()
```
Constructor can probably be removed

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/transport/pipeline/ProtocolHandshaker.java:
```diff
@@ -0,0 +1,188 @@
+
+import static java.lang.String.format;
+
+public class ProtocolHandshaker extends ChannelInboundHandlerAdapter
```
I think this class might benefit from extending `ReplayingDecoder`. It'll then won't have to bother with `handshakeBuffer` and reading exactly 5 ints from the channel.

ali-ince replied:
I checked the `ReplayingDecoder` but found it a bit more complex and cryptic. We'll either need to implement a re-entrant `#decode` method or add some states backed by an `Enum` to track where we left off. Furthermore, we'll still need to cater for left-over bytes in the buffer to send to the newly installed pipeline.

I didn't see much advantage over the current version - which is somewhat not that complex.

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/transport/pipeline/ProtocolHandshaker.java:
```diff
@@ -0,0 +1,188 @@
+                    }
+                    else
+                    {
+                        ctx.writeAndFlush( Unpooled.wrappedBuffer( new byte[]{0, 0, 0, 0} ) ).addListener( ChannelFutureListener.CLOSE );
```
`ctx.alloc()` instead of `Unpooled `

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/runtime/DefaultBoltConnection.java:
```diff
@@ -188,6 +199,12 @@ protected boolean processNextBatch( int batchCount )
                 }
+
+                // we processed all pending messages, let's flush underlying channel
+                if ( queue.size() == 0 || maxBatchSize == 1 )
```
Does this change the previous flush policy?
We used to flush after every message in `MessageProcessingHandler` and now will only flush after all queued messages have been processed. Is my understanding correct?

ali-ince replied:
Previously, `BoltMessagingProtocolHandlerImpl` was keeping track of the number of inbound messages received and number of completed ones ([here](https://github.com/neo4j/neo4j/blob/3.4/community/bolt/src/main/java/org/neo4j/bolt/v1/transport/BoltMessagingProtocolHandlerImpl.java#L118) and [here](https://github.com/neo4j/neo4j/blob/3.4/community/bolt/src/main/java/org/neo4j/bolt/v1/transport/BoltMessagingProtocolHandlerImpl.java#L130)). So, if  `RUN`, `PULLALL` sequence is received, the buffers were flushed only after `PULLALL` processing is finished (given that we don't fill up the internal buffers in `ChunkedOutput`).

We still stick to that behaviour, but now have more control on when to flush.

lutovich commented on community/bolt/src/main/java/org/neo4j/bolt/v1/packstream/ByteBufInput.java:
```diff
@@ -0,0 +1,99 @@
+        buf = requireNonNull( newBuf );
+    }
+
+    public void stop()
```
Can this method be inlined in `#close()`?

ali-ince replied:
I noticed that `Closable` implementation is not needed at all and get rid of it :)

lutovich reviewed (commented):
@ali-ince changes generally look good to me. Made couple comments. Main thing I'm uncertain about is if this PR changes the flushing policy.

ali-ince committed:
Address review comments
